### PR TITLE
Write N-body ICs to a file

### PR DESCRIPTION
**Pull request by marcelo-alvarez**
Adds the ICs class with a method to write particle positions, masses and velocities, suitable for use in N-body sims. See comment below for more details.

### CODE REVIEW

**Comment by marcelo-alvarez** [comment]
@1cosmologist this is ready for you to review (not requesting for you to validate accuracy of the output files).

Running
```
srun -n 16 mockgen mockgen-5 --N 2048 --Niter 1 --laststep writeics --icw --zInit 50
```
on 4 GPU nodes (16 GPUs total) of Perlmutter in the directory
```
/global/cfs/cdirs/mp107d/exgal/users/malvarez/tests/mockgen
```
generates the files
```
/global/cfs/cdirs/mp107d/exgal/users/malvarez/tests/mockgen/output/mockgen-5_13579_Lbox-7700_N-2048_proc[0-15]
```
@jmsexton03 can you have a look and see if these files look ok? If so then we can look at running Nyx on these outside of this issue. They should correspond to the ICs for LCDM cosmology at z=50 in a 7700 Mpc box at 2048^3 resolution, for the following cosmological parameters passed to CAMB:
```
            'T_cmb': 2.72548,
            'h': 0.68,
#=================Omegas========================
            'Omega_m': 0.31,      # Matter density fraction
            'Omega_b': 0.0493890, # baryon density fraction
            'Omega_k': 0.,
#=============Spectral params===================
            'A_s': 2.10058e-9,   # scalar amplitude
            'n_s': 0.96605,      # scalar spectral index
            'alpha_s': 0.,       # scalar spectral running
            'k_pivot': 0.05,     # pivot scale for scalar perturbations
            'r': 0.,             # tensor to scalar ratio
            'n_t': 0.,           # tensor spectral index
            'alpha_t': 0.,       # tensor spectral running
            
#=============reionization========================
            'tau_reion': 0.0543,    # optical depth at reionization
            'YHe': 0.2454,          # Helium fraction
#=============Light relics===========================
            'N_ur': 2.0328,     # Ultra-relativistic species (ultra-light neutrinos)
            'N_ncdm': 1,        # Light massive relics including massive neutrinos
            'm_ncdm': 0.06,     # mass of non-cold light relics
#===========DE EOS===================================
            'w0': -1.,         
            'wa': 0.
```

**Comment by 1cosmologist** [comment]
@marcelo-alvarez Can you add the details of the Nyx output format. It would be useful to have it documented or linked as a part of this PR.

**Comment by marcelo-alvarez** [comment]
@1cosmologist sure here is the documentation for Nyx format: https://amrex-astro.github.io/Nyx/docs_html/ICs.html#start-from-a-binary-file

File: mockgen/ics.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+        rho   = 2.775e11 * omegam * h**2
+        N     = sky.N
+        Lbox  = sky.Lbox
+        z     = sky.zInit
+        a     = 1 / (1+z)
+        Nslab = N // sky.nproc
+        j0    = sky.mpiproc * Nslab
+
+        H = 100 * h * jnp.sqrt(omegam*(1+z)**3+1-omegam) # flat universe with negligible radiation
```
Comment (1cosmologist) [inline mockgen/ics.py:71]: We should be doing this from cosmology module...

File: mockgen/ics.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+        y0 = (j0 + 0.5) * Dgrid     # first point in y-direction (sharding is hardcoded here!)
+        y1 = y0 + (Nslab-1) * Dgrid #  last point in y-direction (sharding is hardcoded here!)
+
+        q1d  = jnp.linspace(x0,x1,N)
+        q1dy = jnp.linspace(y0,y1,Nslab)
+
+        qx, qy, qz = jnp.meshgrid(q1d,q1dy,q1d,indexing='ij')
+
+        x =  qx + D * self.cube.s1x + b0 * D**2 * self.cube.s2x
```
Comment (1cosmologist) [inline mockgen/ics.py:110]: These are already implemented in xgfield. Do we need an alternate implementation?

**Review by 1cosmologist — CHANGES_REQUESTED** [review]
@marcelo-alvarez There seems to be some redundancy with code already implemented in xgfield. I understand not wanting to touch other parts of the pipeline, but for consistent behavior we should not reimplement what already exists, unless absolutely required. Hubble parameter at a certain redshift should be ideally queried from cosmology.

File: mockgen/ics.py (hunk 1 above)
Comment (marcelo-alvarez) [reply mockgen/ics.py:71]: @1cosmologist thanks. Can you please post the line(s) that would replace `H = 100 ...` above such that it is done from the cosmology module?

File: mockgen/ics.py (hunk 2 above)
Comment (marcelo-alvarez) [reply mockgen/ics.py:110]: @1cosmologist thanks for pointing this out. However, I am not aware of how to do this with the code that is in xgfield. Can you post the lines that should replace
```
qx, qy, qz = ...
x = qx + D * ...
```
above? If this would require changes to xgfield, I suggest we consider how to do the refactor in a separate issue / PR.

File: mockgen/ics.py (hunk 1 above)
Comment (1cosmologist) [reply mockgen/ics.py:71]: We would need a simple method in the cosmology module that fetches `Hubble_parameter(z)` from `camb_wsp` (Ref: https://camb.readthedocs.io/en/latest/results.html#camb.results.CAMBdata.hubble_parameter). This should be a simple implementation.

File: mockgen/ics.py (hunk 2 above)
Comment (1cosmologist) [reply mockgen/ics.py:110]: The comoving grid should use the `comoving_q` function in xgfield https://github.com/exgalsky/xgfield/blob/fcbc1db1c9d8fe29ef3cc77c89e72b09d0d7d6a3/xgfield/libfield.py#L54

The euclidean position similarly should use `euclid_i` from xgfield https://github.com/exgalsky/xgfield/blob/fcbc1db1c9d8fe29ef3cc77c89e72b09d0d7d6a3/xgfield/libfield.py#L69

Comment (1cosmologist) [reply mockgen/ics.py:110]: I also suggest moving the velocity implementation to xgfield.

Comment (marcelo-alvarez) [reply mockgen/ics.py:110]: @1cosmologist sounds like we have reached a sufficient number of refactoring changes to address this separately. Can we post a new issue on this in mockgen?

**Comment by marcelo-alvarez** [comment]
See comment inline to review about creating a new issue focused on generic interfaces for this functionality.

File: mockgen/ics.py (hunk 2 above)
Comment (1cosmologist) [reply mockgen/ics.py:110]: I'm not sure we need a new issue here. These codes should not have been a part of writeics in the first place... You can open separate issues in xgfield and cosmology for changes needed there.

**Comment by 1cosmologist** [comment]
Create new issues in xgfield and cosmology to add velocity and Hubble_parameter methods. But these codes should not be a part of mockgen.

**Comment by marcelo-alvarez** [comment]
Thanks @1cosmologist for the review. This code needs to be re-written in order to address the issues you raised, so I'm closing this PR without merging. I will not delete the branch in case some of the code in it can be reused in a new implementation.
